openverse_catalog/dags/providers/provider_api_scripts/inaturalist.py

```diff
@@ -0,0 +1,164 @@
+            (https://wordpress.slack.com/archives/C02012JB00N/p1653145643080479?thread_ts=1653082292.714469&cid=C02012JB00N)
+
+            We use the table structure defined [here,]
+            (https://github.com/inaturalist/inaturalist-open-data/blob/main/Metadata/structure.sql)
+            except for adding ancestry tags to the taxa table.
+"""
+
+import os
+from pathlib import Path
+from typing import Dict
+
+import pendulum
+from airflow.exceptions import AirflowSkipException
+from airflow.operators.python import PythonOperator
+from airflow.providers.amazon.aws.hooks.s3 import S3Hook
+from airflow.providers.postgres.hooks.postgres import PostgresHook
+from airflow.providers.postgres.operators.postgres import PostgresOperator
+from airflow.utils.task_group import TaskGroup
+from common.constants import POSTGRES_CONN_ID
+from common.licenses import LicenseInfo, get_license_info
+from common.loader import provider_details as prov
+from providers.provider_api_scripts.provider_data_ingester import ProviderDataIngester
+
+
+AWS_CONN_ID = os.getenv("AWS_CONN_ID", "test_conn_id")
+PROVIDER = prov.INATURALIST_DEFAULT_PROVIDER
+SCRIPT_DIR = Path(__file__).parents[1] / "provider_csv_load_scripts/inaturalist"
+SOURCE_FILE_NAMES = ["photos", "observations", "taxa", "observers"]
+
+
+class INaturalistDataIngester(ProviderDataIngester):
+
+    providers = {"image": prov.INATURALIST_DEFAULT_PROVIDER}
+
+    def __init__(self, *kwargs):
+        super(INaturalistDataIngester, self).__init__()
+        self.pg = PostgresHook(POSTGRES_CONN_ID)
+
+        # adjustments to buffer limits. TO DO: try to integrate this with the dev
+        # environment logic in the base class, rather than just over-writing it.
+        self.media_stores["image"].buffer_length = 10_000
+        self.batch_limit = 10_000
+        self.sql_template = (SCRIPT_DIR / "export_to_json.template.sql").read_text()
+
+    def get_next_query_params(self, prev_query_params=None, **kwargs):
+        if prev_query_params is None:
+            return {"offset_num": 0}
+        else:
+            next_offset = prev_query_params["offset_num"] + self.batch_limit
+            return {"offset_num": next_offset}
+
+    def get_response_json(self, query_params: Dict):
+        """
+        Call the SQL to pull json from Postgres, where the raw data has been loaded.
+        """
+        sql_string = self.sql_template.format(
+            batch_limit=self.batch_limit, offset_num=query_params["offset_num"]
+        )
+        sql_result = self.pg.get_records(sql_string)
+        # Postgres returns a a list of tuples, even if it's one tuple with one item.
+        return sql_result[0][0]
+
+    def get_batch_data(self, response_json):
+        if response_json:
+            return response_json
+        return None
+
+    def get_record_data(self, data):
+        if data.get("foreign_identifier") is None:
+            return None
```

Comment:
Under what circumstances does this happen? Is it a problem on our end or with the upstream data?

Reply:
What a great question! I actually don't think it does happen because the foreign identifier we're using is the photo id, and inaturalist (almost certainly) has pk constraints on that in their database. But, I think I got the impression that it was something I should test for here just in case. What do you think?

Reply:
I don't think you're wrong to be cautious with assumptions. I wonder if it should be an error condition, or at least something we keep track of when it happens. Especially given how rare we _assume_ it to be, it could be a bad thing if it does happen :slightly_smiling_face: I'd be fine if a follow up issue was created to handle this in an auditable way.

Reply:
Yeah, and actually, :facepalm: there definitely are _not_ pk constraints on the photo table, though I do still think that there are probably not null constraints. _Anyway_, I'll make an issue for this. I also wonder though, how it relates to https://github.com/WordPress/openverse/issues/292, and whether this might be best as part of a more general cross-provider check in the catalog.

Reply:
I'm adding a couple of data quality related follow-ups, actually: https://github.com/WordPress/openverse-catalog/issues/684 and https://github.com/WordPress/openverse-catalog/issues/685 and I will integrate them in the comments.